lib/Analysis/OptimizeRelinearizationAnalysis/OptimizeRelinearizationAnalysis.cpp

```diff
@@ -0,0 +1,370 @@
+    if (llvm::isa<BlockArgument>(value)) {
+      auto type = cast<lwe::RLWECiphertextType>(value.getType());
+      // If the dimension is 3, the key basis is [0, 1, 2] and the degree is 2.
+      int constrainedDegree = type.getRlweParams().getDimension() - 1;
+      MPConstraint *const ct =
+          solver->MakeRowConstraint(constrainedDegree, constrainedDegree, "");
+      ct->SetCoefficient(var, 1);
+    }
+  }
+
+  // For each operation, constrain its inputs to all have the same key basis
+  // degree.
+  std::string cstName;
+  opToRunOn->walk([&](Operation *op) {
+    if (op->getNumOperands() <= 1) {
+      return;
+    }
+
+    std::string name = nameAndLoc(op);
+    auto *anchorVar = keyBasisVars.lookup(op->getOperand(0));
+
+    // degree(operand 0) == degree(operand i)
+    for (OpOperand &opOperand : op->getOpOperands()) {
+      auto *operandDegreeVar = keyBasisVars.lookup(opOperand.get());
+      if (!operandDegreeVar || anchorVar == operandDegreeVar) {
+        continue;
+      }
+      std::stringstream ss;
+      ss << "ArgKeyBasisEquality_" << opOperand.getOperandNumber() << "_"
+         << name;
+      MPConstraint *const ct = solver->MakeRowConstraint(0, 0, ss.str());
+      ct->SetCoefficient(anchorVar, -1);
+      ct->SetCoefficient(operandDegreeVar, 1);
+    }
+  });
+
+  // Some ops require a linear key basis. Return is a special case
+  // where we require returned values from funcs to be linearized.
+  opToRunOn->walk([&](Operation *op) {
+    llvm::TypeSwitch<Operation &>(*op).Case<bgv::RotateOp, func::ReturnOp>(
+        [&](auto op) {
+          for (Value operand : op->getOperands()) {
+            auto *operandDegreeVar = keyBasisVars.lookup(operand);
+            if (!operandDegreeVar) {
+              // This could happen if you return a block argument without doing
+              // anything to it. No variables are created, but it does not
+              // necessarily need to be constrained.
+              if (isa<func::ReturnOp>(op)) return;
+
+              assert(false && "Operand not found in keyBasisVars");
+            }
+            cstName = "RequireLinearized_" + nameAndLoc(op);
+            MPConstraint *const ct = solver->MakeRowConstraint(1, 1, cstName);
+            ct->SetCoefficient(operandDegreeVar, 1);
+          }
+        });
+  });
+
+  // Add constraints that control the effect of relinearization insertion.
+  opToRunOn->walk([&](Operation *op) {
+    llvm::TypeSwitch<Operation &>(*op)
+        .Case<bgv::MulOp>([&](auto op) {
+          // result_degree =
+          //   (arg1_degree + arg2_degree) (1 - insert_relin_op)
+          //   + 1 * insert_relin_op,
+          //
+          // but linearized due to the quadratic term input_key_basis_degree *
+          // insert_relin_op, and an extra variable inserted to keep track of
+          // the difference between the before_relin and after_relin degrees:
+          //
+          // before_relin = arg1_degree + arg2_degree
```

Comment:
-1 here? 

i'm confused how the tests verify with https://github.com/google/heir/blob/bd75f81f4b1d6214498b70a93dd966ea01609237/lib/Dialect/FHEHelpers.h#L24